### PR TITLE
[PM-5531] Improve captcha field autofill disqualification

### PR DESCRIPTION
**Pull request by jprusik**
## Type of change

- Bug fix

## Objective

When gathering/finding login fields to fill, we were qualifying fields (in part) by the values of the field attributes. We would then disqualify _attributes_ based on a constant list of values to ignore (e.g. "captcha, "findanything", etc). However, we weren't disqualifying the whole field, so as long as it qualified otherwise (with a later attribute, for example), we'd use it in autofill.

In Paypal's case, they have a field in their login workflow with an `id` value of "splitPasswordCaptcha" and a `placeholder` value of "Enter code" (what ultimately qualifies the field).

Screenshot 2024-01-16 at 3 36 09 PM

This field would be autofilled with TOTP (when it should be left alone, causing the first step of login to fail with an error (see: https://github.com/bitwarden/clients/issues/5915).

## Code changes

These changes update our logic (in `loadPasswordFields` and `findTotpField`) to properly disqualify the entire field if it has a disqualifying attribute value, so cases like `<input id="captchaforpassword" placeholder="code" />` will no longer autofill.

### CODE REVIEW

File: apps/browser/src/autofill/services/autofill.service.ts [hunk 1]
```diff
@@ -1478,6 +1478,44 @@ export default class AutofillService implements AutofillServiceInterface {
...
+    if (cleanedValue.indexOf("password") < 0) {
+      return false;
+    }
+
+    if (AutoFillConstants.PasswordFieldExcludeList.some((i) => cleanedValue.indexOf(i) > -1)) {
+      return false;
+    }
+
+    return true;
+  }
```
Comment (jprusik) [inline apps/browser/src/autofill/services/autofill.service.ts:1497]: This is just moved out from `loadPasswordFields`

File: apps/browser/src/autofill/services/autofill-constants.ts [hunk 2]
```diff
@@ -44,11 +44,11 @@ export class AutoFillConstants {
 
   static readonly SearchFieldNames: string[] = ["search", "query", "find", "go"];
 
-  static readonly PasswordFieldIgnoreList: string[] = [
+  static readonly FieldIgnoreList: string[] = ["captcha", "findanything", "forgot"];
+
+  static readonly PasswordFieldExcludeList: string[] = [
+    ...AutoFillConstants.FieldIgnoreList,
     "onetimepassword",
```
Comment (jprusik) [inline apps/browser/src/autofill/services/autofill-constants.ts:51]: `onetimepassword` should disqualify password fills, but not `totp`; this should enable us to use disqualifying logic more granularly in the future.

File: apps/browser/src/autofill/services/autofill.service.ts [hunk 3]
```diff
@@ -1496,45 +1534,39 @@ export default class AutofillService implements AutofillServiceInterface {
     fillNewPassword: boolean,
   ) {
     const arr: AutofillField[] = [];
+
     pageDetails.fields.forEach((f) => {
-      if (AutofillService.isExcludedField(f, AutoFillConstants.ExcludedAutofillLoginTypes)) {
+      if (AutofillService.isExcludedFieldType(f, AutoFillConstants.ExcludedAutofillLoginTypes)) {
```
Comment (jprusik) [inline apps/browser/src/autofill/services/autofill.service.ts:1539]: rename for clarity

File: apps/browser/src/autofill/services/autofill.service.ts [hunk 4]
```diff
@@ -1478,6 +1478,44 @@ export default class AutofillService implements AutofillServiceInterface {
...
+    }
+
+    if (AutoFillConstants.PasswordFieldExcludeList.some((i) => cleanedValue.indexOf(i) > -1)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  static fieldHasDisqualifyingAttributeValue(field: AutofillField) {
```
Comment (cagonzalezcs) [inline apps/browser/src/autofill/services/autofill.service.ts:1499]: Can we please add a test that validates this use case when generating the fill data structure?

Comment (jprusik) [reply apps/browser/src/autofill/services/autofill.service.ts:1499]: 21651c083631fdea8348045aa53d6ad2ed05d225

**Review by cagonzalezcs — APPROVED** [review]
Looks good, I saw you already ran the automated tests.

**Comment by jprusik** [comment]
re-running tests post rebase, for safety:
https://github.com/bitwarden/browser-interactions-testing/actions/runs/7587172621

**Review by cagonzalezcs — APPROVED** [review]

**Pull request merged**